### ISSUE TITLE
Excessive memory usage

### ISSUE THREAD

**Issue by beiifeng** [git-for-windows/git#2766]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.27.0.windows.1
cpu: x86_64
built from commit: 907ab1011dce9112700498e034b974ba60f8b407
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [版本 10.0.18363.959]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

** insert your response here **

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

other

![2020-07-28_125409](https://user-images.githubusercontent.com/18087934/88625777-02686500-d0dc-11ea-9e9c-a9c536acba7e.jpg)

Hi, My tools is Intellij IDEA Build #IU-201.8538.31.
when I work with React, I push changes with `nginx.conf`, then I got every large memory use, which about 4G . It takes about 5 minutes to release.

**Comment by derrickstolee** [comment]
@beiifeng: could you copy that full command-line? The image truncates the details. You should also check with Intellij to see what command it is trying to run here that is taking so much memory.

Without knowing _at minimum_ which Git command is being run, we cannot even start to help here.

**Comment by rimrul** [comment]
`git.exe -c credential.helper= -c core.quotepath=false -c log.showSignature=false status --ignored --porcelain -z --` are you sure that's truncated? I think the final `--` is the optional `--` between options and pathspec and then there's no pathspec given.

**Comment by derrickstolee** [comment]
Sorry, the image looked truncated in whatever view I had. My mistake.

**Comment by beiifeng** [comment]
> @beiifeng: could you copy that full command-line? The image truncates the details. You should also check with Intellij to see what command it is trying to run here that is taking so much memory.
> 
> Without knowing _at minimum_ which Git command is being run, we cannot even start to help here.

Ok, I will check with Intellij to find what command Intellij called. Thank you!

**Comment by dscho** [comment]
Does it take as long when you run this command manually, in Git Bash?

**Comment by beiifeng** [comment]
> @beiifeng: could you copy that full command-line? The image truncates the details. You should also check with Intellij to see what command it is trying to run here that is taking so much memory.
> 
> Without knowing _at minimum_ which Git command is being run, we cannot even start to help here.

I take some  screenshot 
1、from Taskmgr

![2020-08-14_113112](https://user-images.githubusercontent.com/18087934/90211269-ceeb4180-de22-11ea-90ce-856285764618.jpg)

2、from Process Explorer
![2020-08-14_113249](https://user-images.githubusercontent.com/18087934/90211390-1eca0880-de23-11ea-8eac-0694dd55617e.jpg)

3、the IDEA install location
![image](https://user-images.githubusercontent.com/18087934/90211456-4e791080-de23-11ea-9d18-7ebec3de6cfc.png)


It happened when I open the React project by IDEA.

**Comment by beiifeng** [comment]
> Does it take as long when you run this command manually, in Git Bash?

YES!

![image](https://user-images.githubusercontent.com/18087934/90212382-c1838680-de25-11ea-885d-1cc92e927c06.png)

this is my node_modules
![image](https://user-images.githubusercontent.com/18087934/90212724-9d747500-de26-11ea-91f2-cd54037a43f4.png)

**Comment by WildCard65** [comment]
Do you by any chance have Windows and GIT configured to use longer paths? If not, then that might explain the warnings.

**Comment by beiifeng** [comment]
> Do you by any chance have Windows and GIT configured to use longer paths? If not, then that might explain the warnings.
> 
> Edit: It appears you don't, Windows by default enforces a path length cap of 260 characters, for legacy reasons, since I think Win 8, it can be configured to allow longer paths.
> 
> Git also has an option to use longer paths as well.

1、I don't change any configure by default.

2、I view the folder  as one sample.
_The screenshot below is from my personal computer, so it looks not as same folder as before_

![image](https://user-images.githubusercontent.com/18087934/90312581-2deeaa80-df38-11ea-8229-bf524fed0143.png)
 **read-pkg-up** in **I:\Rep\crrcdt\base\base-ui\node_modules\_yargs@8.0.2@yargs\node_modules** was a hardlink from **I:\Rep\crrcdt\base\base-ui\node_modules\_read-pkg@2.0.0@read-pkg**
and same as others...
![image](https://user-images.githubusercontent.com/18087934/90312676-de5cae80-df38-11ea-9618-c9b7ad54e4f6.png)

**Comment by beiifeng** [comment]
I got the same result when i run command `git.exe -c credential.helper= -c core.quotepath=false -c log.showSignature=false status --ignored --porcelain -z -- ` on my personal computer .
![20200815203858](https://user-images.githubusercontent.com/18087934/90312742-87a3a480-df39-11ea-89f1-f1627da45a36.jpg)

**Comment by WildCard65** [comment]
Those errors suggest your file paths are longer than 260 characters, the default for Windows.

To allow longer paths, you need to reconfigure Windows and GIT for Windows.

I only know of the group policy method to configure it for the OS, but that may require a PRO license of Windows.

**Comment by beiifeng** [comment]
> Those errors suggest your file paths are longer than 260 characters, the default for Windows.
> 
> To allow longer paths, you need to reconfigure Windows and GIT for Windows.
> 
> I only know of the group policy method to configure it for the OS, but that may require a PRO license of Windows.

1、I set GIT as `git config --global core.longpaths true`  
![20200815221432](https://user-images.githubusercontent.com/18087934/90314632-ac067d80-df47-11ea-9072-6e7e79afff38.jpg)


2、I se OS longpaths enable
![20200815221301](https://user-images.githubusercontent.com/18087934/90314528-0521e180-df47-11ea-983d-6e052036c615.jpg)

3、Restart computer

4、Execute `git.exe -c credential.helper= -c core.quotepath=false -c log.showSignature=false status --ignored --porcelain -z --`

5、Then i got 1G without any warning. but later 4G with another warning.
![20200815221810](https://user-images.githubusercontent.com/18087934/90314643-baed3000-df47-11ea-83cd-b0f39159603c.jpg)

![image](https://user-images.githubusercontent.com/18087934/90314675-056eac80-df48-11ea-8106-519493c75cf1.png)

**Comment by WildCard65** [comment]
Isn't it convention to .gitignore the node modules folder?

**Comment by beiifeng** [comment]
It almost 5G this time, and growth still.

this is my package.json 
```json
{
  "name": "ant-design-pro",
  "version": "1.3.0",
  "description": "An out-of-box UI solution for enterprise applications",
  "private": true,
  "scripts": {
    "precommit": "npm run lint-staged",
    "start": "cross-env BROWSER=none PROXY_TYPE=half ESLINT=none roadhog dev",
    "start:no-mock": "cross-env BROWSER=none PROXY_TYPE=none ESLINT=none roadhog dev",
    "start:all-mock": "cross-env BROWSER=none PROXY_TYPE=all ESLINT=none roadhog dev",
    "build": "cross-env ESLINT=none node --max_old_space_size=4096 node_modules/roadhog/bin/roadhog build",
    "site": "roadhog-api-doc static && gh-pages -d dist",
    "analyze": "cross-env BROWSER=none ANALYZE=true roadhog build",
    "lint:style": "stylelint \"src/**/*.less\" --syntax less",
    "lint:fixstyle": "stylelint \"src/**/*.less\"  --fix",
    "lint": "eslint --ext .js src mock tests && npm run lint:style",
    "lint:fix": "eslint --fix --ext .js src mock tests && npm run lint:style",
    "lint-staged": "lint-staged",
    "lint-staged:js": "eslint --ext .js",
    "test": "roadhog test",
    "test:component": "roadhog test ./src/components",
    "test:all": "node ./tests/run-tests.js",
    "prettier": "prettier --write ./src/**/**/**/*"
  },
  "dependencies": {
    "@ant-design/compatible": "^1.0.1",
    "@antv/data-set": "^0.10.0",
    "@babel/polyfill": "^7.0.0-beta.36",
    "@fullcalendar/core": "^4.4.0",
    "@fullcalendar/daygrid": "^4.4.0",
    "@fullcalendar/interaction": "^4.4.0",
    "@fullcalendar/react": "^4.4.0",
    "@fullcalendar/timegrid": "^4.4.0",
    "@hugojosefson/color-hash": "^2.0.3",
    "@sentry/react": "^5.18.1",
    "antd": "4.0.3",
    "babel-runtime": "^6.9.2",
    "brace": "^0.11.1",
    "classnames": "^2.2.6",
    "crypto-js": "^3.1.9-1",
    "dragm": "0.0.5",
    "driver.js": "^0.3.8",
    "dva": "^2.4.1",
    "dva-loading": "^2.0.3",
    "echarts": "^4.7.0",
    "enquire-js": "^0.2.1",
    "faker": "^4.1.0",
    "gg-editor": "^2.0.4",
    "immutability-helper": "^3.0.0",
    "interactjs": "^1.9.9",
    "intro.js": "^2.9.3",
    "is-type-of": "^1.2.0",
    "jsencrypt": "^3.0.0-rc.1",
    "less": "2.7.2",
    "lodash": "^4.17.11",
    "lodash-decorators": "^6.0.1",
    "moment": "^2.24.0",
    "numeral": "^2.0.6",
    "path-to-regexp": "^3.0.0",
    "prop-types": "^15.6.2",
    "qrcode.react": "^0.9.3",
    "qs": "^6.7.0",
    "randomcolor": "^0.5.4",
    "rc-drawer-menu": "^0.5.0",
    "re-editor": "^0.5.8",
    "react": "^16.8.6",
    "react-beautiful-dnd": "^11.0.3",
    "react-calendar-timeline": "^0.26.7",
    "react-container-query": "^0.11.0",
    "react-copy-to-clipboard": "^5.0.1",
    "react-dnd": "^2.6.0",
    "react-dnd-html5-backend": "^2.5.4",
    "react-document-title": "^2.0.3",
    "react-dom": "^16.8.6",
    "react-file-viewer": "^1.2.1",
    "react-fittext": "^1.0.0",
    "react-highlight-words": "^0.16.0",
    "react-resizable": "^1.7.5",
    "react-rnd": "^10.1.9",
    "react-sortable-hoc": "^1.10.1",
    "react-zmage": "^0.6.5",
    "rollbar": "^2.3.4",
    "slate": "^0.37.3",
    "slate-plain-serializer": "^0.6.2",
    "slate-react": "^0.15.4",
    "slate-schema-violations": "^0.1.32",
    "url-polyfill": "^1.0.10",
    "video-react": "^0.13.2"
  },
  "devDependencies": {
    "babel-eslint": "^8.2.1",
    "babel-plugin-dva-hmr": "^0.4.1",
    "babel-plugin-import": "^1.6.7",
    "babel-plugin-module-resolver": "^3.1.1",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "cross-env": "^5.1.1",
    "cross-port-killer": "^1.0.1",
    "increase-memory-limit": "^1.0.3",
    "enzyme": "^3.1.0",
    "eslint": "^5.0.0",
    "eslint-config-airbnb": "^17.0.0",
    "eslint-config-prettier": "^2.9.0",
    "eslint-plugin-babel": "^5.1.0",
    "eslint-plugin-compat": "^2.1.0",
    "eslint-plugin-import": "^2.8.0",
    "eslint-plugin-jsx-a11y": "^6.0.3",
    "eslint-plugin-markdown": "^1.0.0-beta.6",
    "eslint-plugin-react": "^7.0.1",
    "gh-pages": "^1.0.0",
    "husky": "^3.1.0",
    "lint-staged": "^7.2.0",
    "mockjs": "^1.0.1-beta3",
    "prettier": "1.14.0",
    "pro-download": "^1.0.1",
    "react-loadable": "^5.4.0",
    "redbox-react": "^1.5.0",
    "regenerator-runtime": "^0.12.0",
    "roadhog": "^2.5.0-beta.1",
    "roadhog-api-doc": "^1.1.2",
    "stylelint": "^9.2.1",
    "stylelint-config-prettier": "^3.0.4",
    "stylelint-config-standard": "^18.0.0"
  },
  "optionalDependencies": {
    "puppeteer": "^1.4.0"
  },
  "husky": {
    "hook": {
      "pre-commit": "lint-staged"
    }
  },
  "lint-staged": {
    "(?!color)|**/*.{js,less}": [
      "prettier --write",
      "git add"
    ],
    "**/*.{js}": "lint-staged:js",
    "*.js": [
      "eslint --ext .js",
      "git add"
    ]
  },
  "engines": {
    "node": ">=8.0.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "not ie <= 10"
  ]
}
```

**Comment by beiifeng** [comment]
> Isn't it convention to .gitignore the node_modules folder?

already in

```
# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.

# dependencies
/node_modules
# roadhog-api-doc ignore
/src/utils/request-temp.js
_roadhog-api-doc

# production
/dist

# misc
.DS_Store
npm-debug.log*

/coverage

/.idea

/.svn
debug.log

```

**Comment by dscho** [comment]
I _think_ this might be fixed by https://github.com/git/git/commit/cada7308ade0. Any chance you could

1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
2. `sdk cd git`,
3. `git fetch https://github.com/git/git next && git cherry-pick cada7308ade0`
4. build Git via `make -j$(nproc)`
5. test in-place via `./git --exec-path="$PWD" -C <directory> <command>`?
6. report back

**Comment by beiifeng** [comment]
> I _think_ this might be fixed by [git@cada730](https://github.com/git/git/commit/cada7308ade0). Any chance you could
> 
> 1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
> 2. `sdk cd git`,
> 3. `git fetch https://github.com/git/git next && git cherry-pick cada7308ade0`
> 4. build Git via `make -j$(nproc)`
> 5. test in-place via `./git --exec-path="$PWD" -C <directory> <command>`?
> 6. report back

Thank you !  Because of the Internet, I couldn't download git-sdk before!

I tried it out today：
step 1, step 2, step 3, step 4, and execute command `./git --exec-path="$PWD" -C F:\\Rep\\crrcdt\\base\\base-ui -c credential.helper= -c core.quotepath=false -c log.showSignature=false status --ignored --porcelain -z --` in step 5.

it's Ok with result.
![2020-10-23_183045](https://user-images.githubusercontent.com/18087934/96995449-d3d61b00-1560-11eb-87f5-724cfc802b08.jpg)

![2020-10-23_184048](https://user-images.githubusercontent.com/18087934/96995476-dd5f8300-1560-11eb-8e90-b12de37c1918.jpg)

![2020-10-23_184333](https://user-images.githubusercontent.com/18087934/96995507-e6505480-1560-11eb-8294-db707473bd4a.jpg)

![2020-10-23_184949](https://user-images.githubusercontent.com/18087934/96995547-f5cf9d80-1560-11eb-99ba-5ddc43ea0530.jpg)

**Comment by dscho** [comment]
> it's Ok with result.

Good!
